events/schedulers/autoscaling/autoscaling.go

```diff
@@ -0,0 +1,229 @@
+			autoscaleSettings.ThroughputBucketCount,
+			autoscaleSettings.ThroughputBucketDuration,
+		),
+		autoscaleFrequency: autoscaleSettings.AutoscaleFrequency,
+		autoscalerIn:       make(chan struct{}),
+		autoscalerOut:      make(chan struct{}),
+	}
+
+	for i := 0; i < p.concurrency; i++ {
+		go p.worker()
+	}
+
+	go p.autoscale()
+
+	return p
+}
+
+func (p *Scheduler) Shutdown() {
+	log.Infof("shutting down autoscaling scheduler for %s", p.ident)
+
+	// stop autoscaling
+	p.autoscalerIn <- struct{}{}
+	close(p.autoscalerIn)
+	<-p.autoscalerOut
+
+	log.Info("stopping autoscaling scheduler workers")
+	// stop workers
+	for i := 0; i < p.concurrency; i++ {
+		p.feeder <- &consumerTask{signal: "stop"}
+	}
+	close(p.feeder)
+
+	log.Info("waiting for autoscaling scheduler workers to stop")
+	// wait for all workers to stop
+	for i := 0; i < p.concurrency; i++ {
+		<-p.out
+	}
+	close(p.out)
+
+	log.Info("stopping autoscaling scheduler throughput manager")
+	p.throughputManager.Stop()
+
+	log.Info("autoscaling scheduler shutdown complete")
+}
+
+// Add autoscaling function
+func (p *Scheduler) autoscale() {
+	p.throughputManager.Start()
+	tick := time.NewTicker(p.autoscaleFrequency)
+	defer tick.Stop()
+	for {
+		select {
+		case <-p.autoscalerIn:
+			p.autoscalerOut <- struct{}{}
+			close(p.autoscalerOut)
+			return
+		case <-tick.C:
+			avg := p.throughputManager.AvgThroughput()
+			if avg > float64(p.concurrency) && p.concurrency < p.maxConcurrency {
+				p.concurrency++
+				go p.worker()
+			} else if avg < float64(p.concurrency-1) && p.concurrency > 1 {
+				p.concurrency--
+				p.feeder <- &consumerTask{signal: "stop"}
+			}
+		}
+	}
+}
+
+type consumerTask struct {
+	repo   string
+	val    *events.XRPCStreamEvent
+	signal string
+}
+
+func (p *Scheduler) AddWork(ctx context.Context, repo string, val *events.XRPCStreamEvent) error {
+	p.itemsAdded.Inc()
+	p.throughputManager.Add(1)
+	t := &consumerTask{
+		repo: repo,
+		val:  val,
+	}
+	p.lk.Lock()
+
+	a, ok := p.active[repo]
+	if ok {
+		p.active[repo] = append(a, t)
+		p.lk.Unlock()
+		return nil
+	}
+
+	p.active[repo] = []*consumerTask{}
+	p.lk.Unlock()
+
+	select {
+	case p.feeder <- t:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (p *Scheduler) worker() {
+	log.Infof("starting autoscaling worker for %s", p.ident)
+	p.workersActive.Inc()
+	for work := range p.feeder {
+		for work != nil {
+			// Check if the work item contains a signal to stop the worker.
+			if work.signal == "stop" {
+				log.Infof("stopping autoscaling worker for %s", p.ident)
+				p.workersActive.Dec()
+				p.out <- struct{}{}
+				return
+			}
+
+			p.itemsActive.Inc()
+			if err := p.do(context.TODO(), work.val); err != nil {
+				log.Errorf("event handler failed: %s", err)
+			}
+			p.itemsProcessed.Inc()
+
+			p.lk.Lock()
+			rem, ok := p.active[work.repo]
+			if !ok {
+				log.Errorf("should always have an 'active' entry if a worker is processing a job")
```

Comment:
probably want to continue here

Reply:
oh i guess its technically safe and we do the unlock.. unorthodox but it checks out

Reply:
I think I ripped that straight from the parallel pool just fyi